### PR TITLE
[WIP] fastapi: mount app at root_path instead of prefixing routes and docs URLs

### PR DESCRIPTION
**Pull request by sbidoul**
Since the FastAPI app for each endpoint is an independent app, mounted at a different prefix than the other, I'm wondering if the FastAPI mount mechanism would be more appropriate than prefixing all routers, prefixing the docs urls, etc.

I came to this while debugging the swagger oauth2 redirect url which is currently /docs instead of /{root_path}/docs.

When mounting the app at a prefix, FastAPI prefixes everything correctly.

This PR does not seem to work as is, but I wanted to ask if you had explored this approach @lmignon ?

### CODE REVIEW

File: fastapi/models/fastapi_endpoint.py [hunk 1]
```diff
@@ -200,13 +200,15 @@ def get_uid(self, root_path):
...
-            app.include_router(prefix=self.root_path, router=router)
+            app.include_router(router=router)
         app.dependency_overrides[dependencies.fastapi_endpoint_id] = partial(
             lambda a: a, self.id
         )
         for exception, handler in self._get_app_exception_handlers().items():
             app.add_exception_handler(exception, handler)
-        return app
+        mainapp = FastAPI()
+        mainapp.mount(self.root_path, app)
```
Comment (lmignon) [inline fastapi/models/fastapi_endpoint.py]: Not sure but I think that the path must be unique for each sub app you mount.

**Comment by lmignon** [comment]
Sounds a good idea!

**Comment by lmignon** [comment]
:champagne:

File: fastapi/models/fastapi_endpoint.py (hunk 1 above)
Comment (sbidoul) [reply fastapi/models/fastapi_endpoint.py]: I assume `root_path` is unique across endpoints

File: fastapi/tests/test_fastapi_demo.py [hunk 2]
```diff
@@ -52,7 +54,7 @@ def tearDownClass(cls) -> None:
         super().tearDownClass()
 
     def _get_path(self, path) -> str:
-        return self.fastapi_demo_app.root_path + path
+        return path
```
Comment (sbidoul) [inline fastapi/tests/test_fastapi_demo.py]: But this shows it is a breaking change for tests that rely on `_get_app()` returning prefixed routes.

File: fastapi/tests/test_fastapi_demo.py [hunk 3]
```diff
@@ -17,6 +18,7 @@
 from ..models.fastapi_endpoint_demo import EndpointAppInfo, ExceptionType
 
 
+@odoo.tests.tagged("post_install", "-at_install")
```
Comment (sbidoul) [inline fastapi/tests/test_fastapi_demo.py:19]: And I don't understand why this change suddenly requires tests to run post-install.

File: fastapi/tests/test_fastapi_demo.py (hunk 2 above)
Comment (lmignon) [reply fastapi/tests/test_fastapi_demo.py]: But since we tests the routers and not the app in our addons, it's not a real issue.

File: fastapi/tests/test_fastapi_demo.py (hunk 3 above)
Comment (lmignon) [reply fastapi/tests/test_fastapi_demo.py:19]: It was by chance...

**Review by lmignon — APPROVED** [review]
Works like a charm (Functional tests)

**Review by AnizR — APPROVED** [review]
Great, let's merge?
